go-controller/pkg/network-attach-def-controller/network_manager.go

```diff
@@ -0,0 +1,145 @@
+package networkAttachDefController
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/controller"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+type networkManager interface {
+	EnsureNetwork(util.NetInfo)
+	DeleteNetwork(string)
+	Start() error
+	Stop()
+}
+
+func newNetworkManager(name string, ncm NetworkControllerManager) networkManager {
+	nc := &networkManagerImpl{
+		name:               fmt.Sprintf("[%s network manager]", name),
+		ncm:                ncm,
+		networks:           map[string]util.NetInfo{},
+		networkControllers: map[string]*networkControllerState{},
+	}
+	// this controller does not feed from an informer, networks are manually
+	// added to the queue for processing
+	config := &controller.Config[string]{
+		RateLimiter: workqueue.DefaultControllerRateLimiter(),
+		Reconcile:   nc.sync,
+		Threadiness: 1,
```

Comment:
I would prefer not to introduce new types, I don't they are needed for something that should be fairly simple to understand & doc for now. I feel new types just because of options or tweaks just adds to the confusion.

In regards to the Lister, nothing changed the status quo. You could provide no Lister before and you would have no problem as long as you don't call ReconcileAll. I can add doc and a panic for improper use.

In regards to the ObjNeedsUpdate, the caller of Reconcile can have that logic itself. I can doc.

Reply:
I think documenting every argument of the config and making everyone using it (mostly for controllers with informers) understand which parameters they need to provide and why, is more confusing than just adding a new type. Also things like `controller.Config[string]` may be fine inside the package with a comment, but look strange on the user side. I think making sure the type is used correctly via code is better than doing so via commets + runtime errors.
We can leave Controller interface as is with ReconcileAll panic, but let's at least have an extra config type and hide generic->string inside like
```
type ExternalEventsConfig struct {
	RateLimiter workqueue.RateLimiter
	Reconcile      func(key string) error
	// How many workers should be started for this controller.
	Threadiness int
}

// TODO: explain how this controller works and how it should be used
func NewExternalEventsController(name string, config *ExternalEventsConfig) Controller {
	newConf := &Config[string]{
		RateLimiter: config.RateLimiter,
		Reconcile: config.Reconcile,
		Threadiness: config.Threadiness,
	}
	return NewController[string](name, newConf)
}
```

Reply:
Okay then, let's see where this takes us, hopefully worth it.